parlai/agents/vsepp_caption/modules.py

```diff
@@ -0,0 +1,215 @@
+        if self.max_violation:
+            cost_cap = cost_cap.max(1)[0]
+            cost_im = cost_im.max(0)[0]
+
+        return cost_cap.sum() + cost_im.sum(), ranks, top1
+
+
+class EncoderImage(nn.Module):
+    def __init__(self, embed_size, finetune=False, cnn_type='resnet152',
+                 no_imgnorm=False):
+        """Load pretrained CNN and replace top fc layer."""
+        super().__init__()
+        self.embed_size = embed_size
+        self.no_imgnorm = no_imgnorm
+
+        # Load a pre-trained model
+        self.cnn = self.get_cnn(cnn_type)
+
+        # For efficient memory usage.
+        for param in self.cnn.parameters():
+            param.requires_grad = finetune
+
+        # Replace the last fully connected layer of CNN with a new one
+        if cnn_type.startswith('vgg'):
+            self.fc = nn.Linear(self.cnn.classifier._modules['6'].in_features,
+                                embed_size)
+            self.cnn.classifier = nn.Sequential(
+                *list(self.cnn.classifier.children())[:-1])
+        elif cnn_type.startswith('resnet'):
+            self.fc = nn.Linear(self.cnn.module.fc.in_features, embed_size)
+            self.cnn.module.fc = nn.Sequential()
+
+        self.init_weights()
+
+    def get_cnn(self, arch):
+        """Load a pretrained CNN and parallelize over GPUs
+        """
+        print("=> using pre-trained model '{}'".format(arch))
+        model = models.__dict__[arch](pretrained=True)
+
+        if arch.startswith('alexnet') or arch.startswith('vgg'):
+            model.features = nn.DataParallel(model.features)
+            model.cuda()
+        else:
+            model = nn.DataParallel(model).cuda()
+
+        return model
+
+    def init_weights(self):
+        """Xavier initialization for the fully connected layer
+        """
+        r = np.sqrt(6.) / np.sqrt(self.fc.in_features +
+                                  self.fc.out_features)
+        self.fc.weight.data.uniform_(-r, r)
+        self.fc.bias.data.fill_(0)
+
+    def forward(self, images):
+        """Extract image feature vectors."""
+        features = self.cnn(images)
+        # normalization in the image embedding space
+        features = l2norm(features)
+        # linear projection to the joint embedding space
+        features = self.fc(features)
+        # normalization in the joint embedding space
+        if not self.no_imgnorm:
+            features = l2norm(features)
+
+        return features
+
+
+class EncoderText(nn.Module):
+    def __init__(self, vocab_size, word_dim, embed_size, num_layers,
+                 use_cuda):
+        super().__init__()
+        self.embed_size = embed_size
+        self.use_cuda = use_cuda
+        # word embedding
+        self.embed = nn.Embedding(vocab_size, word_dim)
+        # caption embedding
+        self.rnn = nn.GRU(word_dim, embed_size, num_layers, batch_first=True)
+        self.init_weights()
```

Comment:
this use_cuda stuff is funky here too. when you call enc_text.cuda() in the parent, it pushes all of the submodules here (including the embed and rnn) to the GPU, and the x and lengths should already be cuda tensors (so `I` should already be a cuda tensor)